\documentclass[compress, ucs, xelatex, 11pt, xcolor=x11names, aspectratio=1609,
	hyperref={
		bookmarks=true,
		unicode=true,
		colorlinks=true,
		pdftitle={HybSeq course},
		plainpages=false,
		pdfauthor={Vojtech Zeisek},
		pdfsubject={Practical processing of HybSeq target enrichment sequencing data on computing grids like MetaCentrum},
		pdfcreator={XeLaTeX},
		pdfkeywords={BASH, command line, GNU, HybSeq, Linux, MetaCentrum, sequencing shell, target enrichment},
		linkcolor=Cyan2, % Navigation menu links on pages and in navigation menus
		anchorcolor=Firebrick2, % Not in use?
		citecolor=Firebrick2, % Not in use?
		filecolor=Firebrick2, % Not in use?
		menucolor=Firebrick2, % Not in use?
		urlcolor=Chartreuse2, % Links with \href and \url
		pdftex},
	url={hyphens, lowtilde} % Allow line breaks within URLs
	]{beamer}

% Theme settings
\useoutertheme{infolines}
\useinnertheme[shadow]{rounded}
\usecolortheme{beetle}
% Headline
\setbeamertemplate{headline} {
	\begin{beamercolorbox}{section in head/foot}
		\insertsectionnavigationhorizontal{\paperwidth}{\hskip0pt plus1fill}{\hskip0pt plus1fill}
	\end{beamercolorbox}
	\begin{beamercolorbox}[ht=2ex, dp=1.125ex]{subsection in head/foot}
		\insertsubsectionnavigationhorizontal{\paperwidth}{}{\hfill\hfill}
	\end{beamercolorbox}
	}
% Color background for titles
\setbeamercolor{titlelike}{parent=palette secondary}

% Fonts Linux Libertine
\usepackage{libertine}

% % Other packages
% \usepackage{multicol}
% \usepackage{tabularx}

% % In-line higlighting
\renewcommand{\texttt}[1]{\colorbox{Snow4}{{\ttfamily #1}}}

% Change text color of highlighted text
\renewcommand{\alert}[1]{\textcolor{OrangeRed2}{#1}}

% Syntax higlight
\usepackage{minted}
\usemintedstyle{vim} % Styles are listed by pygmentize -L styles; languages are listed by pygmentize -L lexers
\newminted{bash}{linenos, fontsize=\footnotesize, bgcolor=Snow4, fontfamily=tt, gobble=4, numbersep=-3pt}
\newminted{splus}{linenos, fontsize=\footnotesize, bgcolor=Snow4, fontfamily=tt, gobble=4, numbersep=-3pt}
% Change line number style
\renewcommand{\theFancyVerbLine}{
	\sffamily
	\textcolor{DarkOrchid4}{
		\tiny
		\oldstylenums{
			\arabic{FancyVerbLine}
			}
		}
	}

% Default language
\usepackage[main=american]{babel}

% Quotes
	\usepackage[autostyle=true, english=american]{csquotes}

% Title page
\author{Vojtěch Zeisek}
\institute[\url{https://trapa.cz/}]{Department of Botany, Faculty of Science, Charles University, Prague\\Institute of Botany, Czech Academy of Sciences, Průhonice\\\url{https://trapa.cz/}, \href{mailto:zeisek@natur.cuni.cz}{zeisek@natur.cuni.cz}}
\title{HybSeq course}
\subtitle{Practical processing of HybSeq target enrichment sequencing data on computing grids like MetaCentrum}
\titlegraphic{\includegraphics[width=1.5cm]{konsole.png}}
\date{January 20 to 23, 2020}

\begin{document}

\begin{frame}
	\titlepage
\end{frame}

\begin{frame}[allowframebreaks]{Outline}
	\tableofcontents
\end{frame}

\section{Introduction}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Test data}

\begin{frame}{\textit{Oxalis} test data set}
	\begin{itemize}
		\item Genus \textit{Oxalis} has altogether ca. 500 species, over 200 in South Africa
		\item We selected 24 South African species (following \href{https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.12487}{Schmickl et al. 2016}) as test data
		\item The probes used for sequencing were introduced in \href{https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.12487}{Schmickl et al. 2016}, but were reduced (some probes with poor sequencing results were removed)
	\end{itemize}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Data processing overview}

\begin{frame}[allowframebreaks]{Steps from sequencing files to species trees}
	\begin{enumerate}
		\item Trimming of raw sequencing FASTQ files (removal of adaptors,~\ldots)
		\item Deduplication of FASTQ reads
		\begin{itemize}
			\item Not strictly required, duplicates mainly provide wrong insight into real coverage of particular loci
		\end{itemize}
		\item Checking of FASTQ files in FastQC or similar tool and removal of low-quality files
		\item Preparing probe reference FASTA file and list of samples for processing by HybPiper
		\item Processing every sample with with \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper} (or \href{https://github.com/tomas-fer/HybPhyloMaker}{HybPhyloMaker} or similar tool)
		\begin{enumerate}
			\item Mapping of FASTQ reads with \href{https://github.com/lh3/bwa}{BWA} to FASTA reference
			\item Distributing (sorting) of reads according to successful hits (using \href{https://github.com/samtools/samtools}{Samtools}) into FASTA files for assembly
			\item Assembly of sorted reads with \href{https://github.com/ablab/spades}{SPAdes}
			\item Alignment of SPAdes contigs against the target sequence
			\begin{itemize}
				\item Contigs are not expected to overlap much
				\item Initial exonerate search is filtered for hits that are above a certain threshold
				\item Contigs that pass this filter are arranged in order along the alignment
				\item All contigs that pass the previous steps are concatenated into a \enquote{supercontig} and the exonerate search is repeated
			\end{itemize}
			\item Search for paralogs --- if SPAdes assembler generates multiple contigs that contain coding sequences representing 75\% of the length of the reference protein, HybPiper will print a warning for that gene
			\item Recovering of the individual sequences
			\item Statistics of the recovery
			\item Cleanup of temporal files (especially SPAdes produces huge amount of data unneeded for further processing)
		\end{enumerate}
		\item Statistics of sequence lengths in all samples and more information about recovered contigs
		\item Creation of heatmaps (using \href{https://www.r-project.org/}{R} and packages \href{https://cran.r-project.org/package=gplots}{gplots} and \href{https://cran.r-project.org/package=heatmap.plus}{heatmap.plus}, or \href{https://cran.r-project.org/package=ggplot2}{ggplot2} and \href{https://cran.r-project.org/package=reshape2}{reshape2})
		\item Retrieve of sequences of exons, introns and supercontigs for all samples
		\item Alignment of all contigs (e.g. by \href{https://mafft.cbrc.jp/alignment/software/}{MAFFT}; or \href{https://www.drive5.com/muscle/}{MUSCLE}, \href{http://www.clustal.org/}{Clustal},~\ldots{ }e.g. using \href{https://www.r-project.org/}{R} and packages \href{https://cran.r-project.org/package=ape}{ape} and/or \href{https://cran.r-project.org/package=ips}{ips})
		\begin{itemize}
			\item All alignments must be trimmed --- columns/rows with too much missing data (e.g. beginning and end of the alignment) must be removed (e.g. using \href{https://www.r-project.org/}{R} and package \href{https://cran.r-project.org/package=ape}{ape})
			\item It is also useful to create simple NJ tree graphical check of alignment (e.g. using \href{https://www.r-project.org/}{R} and package \href{https://cran.r-project.org/package=ape}{ape})
		\end{itemize}
		\item Sorting of alignments, statistics of their length and quality, discarding of poor (too short, too few individuals, too few variable positions,~\ldots) alignments
		\item Reconstruction of gene trees from all aligned contigs (e.g. using \href{http://www.iqtree.org/}{IQ-TREE}, or \href{https://github.com/stamatak/ExaML}{ExaML}, \href{https://nbisweden.github.io/MrBayes/}{MrBayes}, \href{http://www.atgc-montpellier.fr/phyml/}{PhyML}, \href{https://github.com/stamatak/standard-RAxML}{RAxML},~\ldots)
		\item Post-processing of gene trees
		\begin{itemize}
			\item Identification, inspection and possible removal of gene trees with significantly different topology
			\item 
		\end{itemize}
		\item 
	\end{enumerate}
\end{frame}

\subsection{Software needed}

\begin{frame}[allowframebreaks]{List of software used during the course}
	\begin{itemize}
		\item \href{https://github.com/smirarab/ASTRAL}{ASTRAL} (see lesson by TF)
		\item BASH 4 or later and GNU core utils (\enquote{Linux command line})
		\item \href{ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/}{BLAST+} (used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper})
		\item \href{https://github.com/lh3/bwa}{BWA} (used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper})
		\item \href{https://www.ebi.ac.uk/about/vertebrate-genomics/software/exonerate}{Exonerate} (used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper})
		\item \href{https://www.gnu.org/software/parallel/}{GNU Parallel} (used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper} and in BASH scripts)
		\item \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper}
		\item \href{http://www.iqtree.org/}{IQ-TREE}
		\item \href{https://mafft.cbrc.jp/alignment/software/}{MAFFT}
		\item \href{https://github.com/mossmatters/MJPythonNotebooks}{MJPythonNotebooks} (used by \href{https://bitbucket.org/blackrim/phyparts}{phyparts})
		\item \href{https://bioinfocs.rice.edu/PhyloNet}{PhyloNet}
		\item \href{https://bitbucket.org/blackrim/phyparts}{phyparts}
		\item \href{https://www.python.org/}{Python 2.7 or later} and \href{https://biopython.org/}{Biopython 1.59 or later} (used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper})
		\item \href{https://github.com/lutteropp/QuartetScores}{QuartetScores} (see lesson by TF)
		\item \href{https://www.r-project.org/}{R 3.5 or later} and packages \href{https://cran.r-project.org/package=ade4}{ade4}, \href{https://cran.r-project.org/package=adegenet}{adegenet}, \href{https://cran.r-project.org/package=ape}{ape}, \href{https://cran.r-project.org/package=corrplot}{corrplot}, \href{https://cran.r-project.org/package=distory}{distory}, \href{https://cran.r-project.org/package=ggplot2}{ggplot2}, \href{https://cran.r-project.org/package=gplots}{gplots}, \href{https://cran.r-project.org/package=heatmap.plus}{heatmap.plus}, \href{https://cran.r-project.org/package=ips}{ips}, \href{https://cran.r-project.org/package=kdetrees}{kdetrees}, \href{https://cran.r-project.org/package=pegas}{pegas}, \href{https://cran.r-project.org/package=phangorn}{phangorn}, \href{https://cran.r-project.org/package=phytools}{phytools} and \href{https://cran.r-project.org/package=reshape2}{reshape2}
		\begin{itemize}
			\item Used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper}, for alignment of contigs, post-processing of alignments, post-processing and comparison of gene trees, etc.
		\end{itemize}
		\item \href{https://github.com/samtools/samtools}{Samtools} (used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper})
		\item \href{https://github.com/ablab/spades}{SPAdes} (used by \href{https://github.com/mossmatters/HybPiper/wiki}{HybPiper})
		\item \href{https://github.com/uym2/TreeShrink}{TreeShrink}
	\end{itemize}
\end{frame}

\subsection{MetaCentrum computing environment}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\section{Data preprocessing}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{General data structure}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Trimming}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Deduplication}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Preparing data for HybPiper}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\section{HybPiper}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Preparation}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Processing input files}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Retrieving sequences}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\section{Alignments}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Sorting alignments}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\section{Gene trees}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Post-processing gene trees}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Post-processing gene trees}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\section{Comparing gene trees}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Filtering trees}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Visualizing differences among trees}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\subsection{Phylogenetic networks}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\section{The end}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\begin{frame}[fragile]{}
	\begin{itemize}
		\item 
	\end{itemize}
	\begin{spluscode}
    
	\end{spluscode}
	\begin{bashcode}
    
	\end{bashcode}
\end{frame}

\end{document}
